js/recipe_manager.js

```diff
@@ -1,84 +1,119 @@
-    "total_calories": 2600,
-    "number_of_portions": 4,
-    "calories_per_portion": 650
+    'instructions': '1 Heat pasta water. Prepare rest of ingredients. Voila.',
+    'total_calories': 2600,
+    'number_of_portions': 4,
+    'calories_per_portion': 650
   },
-  { 
-    "id": 2,
-    "title": "Toast",
-    "ingredients": [
-      "bread",
-      "cheese"
+  {
+    'id': 2,
+    'title': 'Toast',
+    'ingredients': [
+      'bread',
+      'cheese'
     ],
-    "instructions": "Put cheese beetwen bread slices. Put in the toaster. Voila",
-    "total_calories": 800,
-    "number_of_portions": 2,
-    "calories_per_portion": 400
+    'instructions': 'Put cheese beetwen bread slices. Put in the toaster. Voila',
+    'total_calories': 800,
+    'number_of_portions': 2,
+    'calories_per_portion': 400
   }
   ]
 }
 
-
 // ----------------- API -------------------
 
-const validateRecipeKey = function(key, newValue) {
-  // total_calories must be a number
-  // number_of_portions must be a number
-};
-
-const validateRecipe = function(recipe) {
-  // loop over pairs: key-value, call the function validateRecipeKey() with key, value
-};
-
 // CRUD
 
 /** Returns all recipes in the database */
-const getRecipes = function() {};
+const getRecipes = function () {
+  return window.db.recipes
+}
 
 /** Returns recipe with the given id */
-const getRecipe = function(id) {};
+const getRecipe = function (id) {
+  return window.db.recipes.find(function (el) {
+    return el.id === id
+  })
+}
 
 /** Adds new recipe to the database. Returns the new recipe's ID */
-const createRecipe = function(recipe) {
-  validateRecipe(recipe);
-  const numOfPort = recipe.number_of_portions;
-  const totalCal = recipe.total_calories;
- 
-  // TODO: calculate calories per portion;
-  // TODO: generate new ID;
+const createRecipe = function (recipe) {
+  if (!validateRecipe(recipe)) {
+    return false
+  };
+
+  const numOfPort = recipe.number_of_portions
+  const totalCal = recipe.total_calories
+  const caloriesPerPortion = totalCal / numOfPort
+  const newId = uuidv1()
+  const myRecipe = {
+    id: newId,
+    title: recipe.title,
+    ingredients: recipe.ingredients,
+    instructions: recipe.instructions,
+    total_calories: recipe.total_calories,
+    number_of_portions: recipe.number_of_portions,
+    calories_per_portion: caloriesPerPortion
+  }
+  window.db.recipes.push(myRecipe)
 
-  return ID;
-};
+  return newId
+}
 
 /** Updates recipe in the database.  */
-const updateRecipe = function(id, key, newValue) {
-  validateRecipeKey(key, newValue) 
+const updateRecipe = function (id, key, newValue) {
+  if (!validateRecipeKey(key, newValue)) {
+    if (!isValidKey(key)) {
+      throw new Error("Your key doesn't exist")
+    } else {
+      return false
+    }
+  };
+
+  const doesIdExist = window.db.recipes.find(function (el) {
+    return el.id === id
+  })
+
+  if (!doesIdExist) {
+    throw new Error("Your id doesn't exist")
```

Comment:
This particular error on the other hand can stay here as it relies on the database. Alternatively, to keep the validations namespace pure, you could move it into a function to which you pass the whole database and perform the same check.